[WIP] WebSocket active scan infrastructure

eakirtas opened the pull request:
This pull request its part of WebSocket Active Scan project which described in issue #4276.
___________________________________________________________
**[WS-Active-Scan] WebSocket active scan infrastructure**

_(on early stage yet)_

This commit introduces a new active scanner feature on WebSocket add-on. The WebSocket scanner infrastructure based on WebSocket Treemap and Structure analyzer which are introduced in zaproxy/zap-extensions #1742.

**Short Intro**

WebSocket Protocol comes with many differences in reference to Http protocol. The most common one is the format of the messages/payloads. In contrast with HTTP, WebSocket protocol does not define the specific format of the messages will be transmitted over the TCP connection. That means at the same connection may be transferred messages with a different format. In addition, WebSocket protocol keeps the TCP connection opened in order to offer us a full duplex client/server communication in contrast with the classic request/response technique. Those are the most challenging parts which are tried to overcome by this PR. In any other case, this implementation attempts to emulate the existing HTTP active scanner.

**The concept**

There as some utility classes which are managed the whole scanning process.

- `WebSocketActiveScanManager`: The class which manage the scans and links the active scanner with the rest functionality of WebSocket Add-on. This class can be used in order to add a new `WebSocketActivePlugin` (Explained above) and start a new Active Scan. Note: The scan will not run if there is another scan was already running.

- `WebSocketPluginFactory`: Used for storing and loading plugins.

- `WebSocketTarget`: Stores the basic information about the active scan targets.

- `WebSocketScanner`: Implements the runnable class. This class is responsible to separate the targets onto different host process (if there are more than two). That could be happened when the user selects to scan all nodes under the root. In addition, if there are available threads `WebSocketScanner` can create different tasks for those different hosts in order to run scans in parallel. Finally, If the user selects a leaf node (ex. `MessageNode`) the `WebSocketScanner` will create a new HostProcess in order to handle the communication with the server.

- `WebSocketHostProcess`: The host process represents a scan on a specific host. It has to distribute the targets over the appropriate plugins. In addition, `WebSocketHostProcess` offers to the plugins some basic utilities. Those utilities help plugins to establish a new WebSocket Connection, send messages, receive the response messages from the appropriate channel etc. I should note that every plugin has to open a new channel, so every time a new message received the WebSocketHostProcess be able to inform the appropriate plugin. The `WebSocketScanner` has many host process (which could be running on a different thread).

All those classes aim to support the scanning plugins. Every plugin have to test the application (either on server site or client site) in order to find a set of vulnerabilities. To simplify the whole process I create some abstracts class (as the HTTP active scanner does). In order to write a new plugin, we should extend the following classes. An important note: The plugins are categorized in such a way in order to test the different "parts" of the WebSocket Connection. This means that a plugin must scan some specific node's type. That depends on the class you extend. Finally, the target is defined by "user" (either from UI or API). Target could be either recursive or not. The target (which reached to the plugin) could be any available (except folder) under the HostFolder. `The WebSocketActivePlugin` is the class which is responsible to iterate over the nodes.

- `WebSocketActivePlugin`: This is the superclass of all plugin. You typically will not extend this class directly but provides key methods that you will need to use and abstract methods that will need to be implemented. (exactly as `AbstractPlugin`). `WebSocketActivePlugin` iterates over the targets, "ask" the scanner if it is interested to scan this target (by `applyScan` method) and then gives them the target to scan it. Implement `boolean applyScan(StructuralWebSocketNode structuralWebSocketNode)` in order to choose which nodes you to apply scan and `void canNode(StructuralWebSocketNode structuralWebSocketNode)` to scan it. It is able to apply scans on nodes: {`HANDSHAKE`, `MESSAGE_OBJECT`, `MESSAGE_TEXT`}

- `WebSocketActiveHandshakeScan`: Extend this class in order to perform some tests on the handshake process. Implement `boolean applyScanToHandshake(HttpMessage handshake)` in order to choose which handshakes you want to scan. And of course, implement `scanHandshake(HttpMessage handshake)` in order to scan the handshake. `WebSocketActivePlugin` is able to apply scans on nodes with type: {`HANDSHAKE`}

- `WebSocketMessageNodeScan` : Extend this class in order to perform scans on nodes. `WebSocketMessageNodes` stores the `WebSocketMessages`, `PayloadAnalyzers`, `PayloadStructure` etc. Implement `boolean applyScan(WebSocketMessageNode messageNode)` in order to choose in which nodes you want to apply your scan and `void scanMessageNode(WebSocketMessageNode messageNode)`  to scan it. It is able to apply scans on nodes with type: {`MESSAGE_OBJECT`, `MESSAGE_TEXT`}

- `WebSocketMessageDTOScan`: Extend this class in order to perform a scan on Messages. Implement `boolean applyScan(WebSocketMessageDTO message)` in order to choose in which Messages you want to apply your scan and `scanMessage(WebSocketMessageDTO message)` to scan it. It is able to apply scans on nodes with type: {`MESSAGE_OBJECT`, `MESSAGE_TEXT`}

- `WebSocketPayloadScan`: Extend this class in order to perform scans on Payloads. Implement `boolean applyScan(Object message)` in order to choose in which and `scanPayload(Object message)` to scan it. It is able to apply scans on nodes: {`MESSAGE_OBJECT`,` MESSAGE_TEXT`}

- `WebSocketStructureScan`: Extend this class in order to perform scans on Paylaod Structure. This scan based on Payload Structure which introduced on zaproxy/ zap-extensions #1742. Payload structure exported from `PayloadAnalyzers`. Those interfaces aim to analyze the WebSocket message in a structural format (expressed as a data structure). You can implement those in order to adjust it in your test case. The basic idea behind that is to create a formal structure in order to iterate over the different parameter (keys and values). If your messages can be mapped in a `PayloadStructure` you can use it to scan different parameters and create your own attacks vectors with analyzers. (In an abstract way `PayloadAnalyzers` has the role to deserialize/serialize the payloads). Implement `boolean applyScan(PayloadAnalyzer analyzer, PayloadStructure structure)` in order to choose in which and `scanMessageStructure(PayloadAnalyzer analyzer, PayloadStructure structure)` to scan it. `PayloadAnalyzer` and `PayloadStructure` should not be null. It is able to apply scans on nodes with type: {`MESSAGE_OBJECT`, `MESSAGE_TEXT`}

- `WebSocketMessageParamScan`: Extend this class in order to perform scans on parameters. This plugin scanning a target parameter by parameter. `PayloadAnalyzer` and `PayloadStructure` should not be null. It is able to apply scans on nodes with type: {`MESSAGE_OBJECT`, `MESSAGE_TEXT`}.

All Plugins have to implement and the following methods:

- `void messageReceived(WebSocketMessageDTO message)` : Notify the plugin that a new message received over the channel
- `void connectionStateChanged(WebSocketProxy.State state, WebSocketProxy proxy)`:State of connection changed
- `String getName()` Return the name of the plugin
- `int getCode()` returns the unique ID of the plugin

**Plugins**

- `ScriptWebSocketMessageActivePlugin`: This class extends WebSocketMessageNodeScan and uses the ExtensionScript. Calls the script is written by the user. Implement in your script the WebSocket ActiveMessageScript interface in order to create an active scan.

- `WebSocketActiveMessageScript`: Should be implemented by active scan scripts
  - [Optional] `boolean applyScan(WebSocketMessageNode webSocketMessageNode) `return true by default
  - `scan(WebSocketMessageNode webSocketMessageNode, ScriptWebSocketMessageActivePlugin parent)`;
  - `void messageReceived(WebSocketMessageDTO message)`
  - [Optional]` void connectionStateChanged(WebSocketProxy.State state)`

- `WebsocketActiveScan `Template.js is a template script

**User Interface**

Gives the ability to user to select the starting node.

**Issues**
- [ ] Address the concurrency issues
- [ ] ....

**TODO**
- [ ] Clean up the code 
- [ ] Add documentation
- [ ] ...

_(Not ready for code Review yet)_

eakirtas committed:
[WS-Tree-Map] WebSocket Tree Map Structure

Introducing to new WebSocket Tree:

* Nodes: All WebSocket Nodes should implement the interface
  StructuralWebSocketNode. That interface helps to store the tree structure
  on a database in the future.

  ** WebSocketTreeNode The base class of all Node, it is an abstract class
     which offers the basic method to create the nodes and navigate over them.

  ** WebSocketNodeType: Helps to define the different use of the node. The
     different types maybe will act differently. (ex. the FOLDER_HANDSHAKES
     keeps the handshakes of a connection, MESSAGE_PONG it is a node which
     keeps ping messages).

  ** WebSocketFolderNode: This nodes acts as a node folder which keeps a
     specific kind of child. (ex FOLDER_HOST always keeps the appropriate
     folders for a host, FOLDER_HANDSHAKES keeps the paths of the handshakes)

  ** WebSocketMessageNode: Keeps the WebSocket Messages and the appropriate
     stuff to analyze the messages.  WebSocket Messages represented as DTOs.

  ** WebSocketHandshakeNode: Keeps the http handshake.

  All channels has the same following structure:

  |- ROOT_FOLDER
      |- HOST_FOLDER
         |- HANDSHAKE_FOLDER
            |- <path> HANDSHAKE_FOLDER
               |- <to> HANDSHAKE_FOLDER
                  |- <Handshake> HANDSHAKE
         |- HEARTBEAT_FOLDER
            |- <Ping Message> PING_MESSAGE
            |- <Pong Message> PING_MESSAGE
         |- MESSAGES_FOLDER
            |- (Analyzed Payload Structure) MESSAGE
         |- CLOSE_FOLDER
            |- <Close Message> CLOSE_MESSAGE

* WebSocketMap:
  Manage whole functionality of the tree. Add connections and Messages to
  appropriate position in the tree. Initialize a new tree by static method
  createTree.

TODO:
 [ ] Correct the following: path to the handshake represented as
  Handshake type not as HandshakeFolders
 [ ] Implement UI
 [ ] Implement API
 [ ] Add documentation
 [ ] Extend the existing interface StructuralWebSocketNode
 [ ] Sort folders under HOST_FOLDER

eakirtas committed:
[WS-Tree-Map] Tests for node and the map

eakirtas committed:
[WS-Tree-Map] Test For Payload Analyzer

eakirtas committed:
[WS-Tree-Map] Basic UI

eakirtas committed:
[WS-Tree-Map] Some changes to Adapt Structural Tree to UI

[WS-Tree-Map] Add History references & add some constants

eakirtas committed:
[WS-Active-Scan] WebSocket active scan infrastructure

(on early stage yet)

This commit introduces a new active scanner feature on WebSocket add-on. The WebSocker scanner infrastructure based on WebSocket Treemap and Structure analyzer which are introduced in zaproxy/ zap-extensions #1742.

**Short Intro**

WebSocket Protocol comes with many differences in reference to Http protocol. The most common one is the format of the messages/payloads. In contrast with HTTP, WebSocket protocol does not define the specific format of the messages will be transmitted over the TCP connection. That means at the same connection may be transferred messages with a different format. In addition, WebSocket protocol keeps the TCP connection opened in order to offer us a full duplex client/server communication in contrast with the classic request/response technique. Those are the most challenging parts which are tried to overcome by this PR. In any other case, this implementation attempts to emulate the existing HTTP active scanner.

**The concept**

There as some utility classes which are managed the whole scanning process.

- WebSocketActiveScanManager: The class which manage the scans and links the active scanner with the rest functionality of WebSocket Add-on. This class can be used in order to add a new WebSocketActivePlugin (Explained above) and start a new Active Scan. Note: The scan will not run if there is another scan was already running.

- WebSocketPluginFactory: Used for storing and loading plugins.

- WebSocketTarget: Stores the basic information about the active scan targets.

- WebSocketScanner: Implements the runnable class. This class is responsible to separate the targets onto different host process (if there are more than two). That could be happened when the user selects to scan all nodes under the root. In addition, if there are available threads WebSocketScanner can create different tasks for those different hosts in order to run scans in parallel. Finally, If the user selects a leaf node (ex. MessageNode) the WebSocketScanner will create a new HostProcess in order to handle the communication with the server.

- WebSocketHostProcess: The host process represents a scan on a specific host. It has to distribute the targets over the appropriate plugins. In addition, WebSocketHostProcess offers to the plugins some basic utilities. Those utilities help plugins to establish a new WebSocket Connection, send messages, receive the response messages from the appropriate channel etc. I should note that every plugin has to open a new channel, so every time a new message received the WebSocketHostProcess be able to inform the appropriate plugin. The WebSocketScanner has many host process (which could be running on a different thread).

All those classes aim to support the scanning plugins. Every plugin have to test the application (either on server site or client site) in order to find a set of vulnerabilities. To simplify the whole process I create some abstracts class (as the HTTP active scanner does). In order to write a new plugin, we should extend the following classes. An important note: The plugins are categorized in such a way in order to test the different "parts" of the WebSocket Connection. This means that a plugin must scan some specific node's type. That depends on the class you extend. Finally, the target is defined by "user" (either from UI or API). Target could be either recursive or not. The target (which reached to the plugin) could be any available (except folder) under the HostFolder. The WebSocketActivePlugin is the class which is responsible to iterate over the nodes.

- WebSocketActivePlugin: This is the superclass of all plugin. You typically will not extend this class directly but provides key methods that you will need to use and abstract methods that will need to be implemented. (exactly as AbstractPlugin). WebSocketActivePlugin iterates over the targets, "ask" the scanner if it is interested to scan this target (by applyScan method) and then gives them the target to scan it. Implement <boolean applyScan(StructuralWebSocketNode structuralWebSocketNode)> in order to choose which nodes you to apply scan and <void canNode(StructuralWebSocketNode structuralWebSocketNode)> to scan it. It is able to apply scans on nodes: {HANDSHAKE, MESSAGE_OBJECT, MESSAGE_TEXT}

- WebSocketActiveHandshakeScan: Extend this class in order to perform some tests on the handshake process. Implement <boolean applyScanToHandshake(HttpMessage handshake)> in order to choose which handshakes you want to scan. And of course, implement scanHandshake(HttpMessage handshake) in order to scan the handshake. WebSocketActivePlugin is able to apply scans on nodes with type: {HANDSHAKE}

- WebSocketMessageNodeScan : Extend this class in order to perform scans on nodes. WebSocketMessageNodes stores the WebSocketMessages, PayloadAnalyzers, PayloadStructure etc. Implement <boolean applyScan(WebSocketMessageNode messageNode)> in order to choose in which nodes you want to apply your scan and <void scanMessageNode(WebSocketMessageNode messageNode)>  to scan it. It is able to apply scans on nodes with type: {MESSAGE_OBJECT, MESSAGE_TEXT}

- WebSocketMessageDTOScan: Extend this class in order to perform a scan on Messages. Implement <boolean applyScan(WebSocketMessageDTO message)> in order to choose in which Messages you want to apply your scan and <scanMessage(WebSocketMessageDTO message)> to scan it. It is able to apply scans on nodes with type: {MESSAGE_OBJECT, MESSAGE_TEXT}

- WebSocketPayloadScan: Extend this class in order to perform scans on Payloads. Implement <boolean applyScan(Object message)> in order to choose in which and <scanPayload(Object message)> to scan it. It is able to apply scans on nodes: {MESSAGE_OBJECT, MESSAGE_TEXT}

- WebSocketStructureScan: Extend this class in order to perform scans on Paylaod Structure. This scan based on Payload Structure which introduced on zaproxy/ zap-extensions #1742. Payload structure exported from PayloadAnalyzers. Those interfaces aim to analyze the WebSocket message in a structural format (expressed as a data structure). You can implement those in order to adjust it in your test case. The basic idea behind that is to create a formal structure in order to iterate over the different parameter (keys and values). If your messages can be mapped in a PayloadStructure you can use it to scan different parameters and create your own attacks vectors with analyzers. (In an abstract way PayloadAnalyzers has the role to deserialize/serialize the payloads). Implement boolean applyScan(PayloadAnalyzer analyzer, PayloadStructure structure)> in order to choose in which and scanMessageStructure(PayloadAnalyzer analyzer, PayloadStructure structure)> to scan it. PayloadAnalyzer and PayloadStructure should not be null. It is able to apply scans on nodes with type: {MESSAGE_OBJECT, MESSAGE_TEXT}

- WebSocketMessageParamScan: Extend this class in order to perform scans on parameters. This plugin scanning a target parameter by parameter. PayloadAnalyzer and PayloadStructure should not be null. It is able to apply scans on nodes with type: {MESSAGE_OBJECT, MESSAGE_TEXT}.

All Plugins have to implement and the following methods:

- void messageReceived(WebSocketMessageDTO message) Notify the plugin that a new message received over the channel
- void connectionStateChanged(WebSocketProxy.State state, WebSocketProxy proxy) State of connection changed
- String getName() Return the name of the plugin
- int getCode() returns the unique ID of the plugin

**Plugins**

- ScriptWebSocketMessageActivePlugin: This class extends WebSocketMessageNodeScan and uses the ExtensionScript. Calls the script is written by the user. Implement in your script the WebSocket ActiveMessageScript interface in order to create an active scan.

- WebSocketActiveMessageScript: Should be implemented by active scan scripts
  - [Optional] boolean applyScan(WebSocketMessageNode webSocketMessageNode) return true by default
  - scan(WebSocketMessageNode webSocketMessageNode, ScriptWebSocketMessageActivePlugin parent);
  - void messageReceived(WebSocketMessageDTO message);
  - [Optional] void connectionStateChanged(WebSocketProxy.State state);

- WebsocketActiveScan Template.js is a template script

**User Interface**

Gives the ability to user to select the starting node.

psiinon commented:
This pull request **introduces 9 alerts** when merging 9da329a6057d94debfffee336f445daa7aabbfc3 into b33ac46d6bb7db10ed20d53a51ff48dd8d3010de - [view on LGTM.com](https://lgtm.com/projects/g/zaproxy/zap-extensions/rev/pr-ad8ced91644d1145dbb0bd3ca9e11676623d3a3e)

**new alerts:**

* 2 for Equals on incomparable types
* 2 for Implicit conversion from array to string
* 1 for Container contents are never accessed
* 1 for Inconsistent equals and hashCode
* 1 for Type mismatch on container modification
* 1 for Reference equality test of boxed types
* 1 for Dereferenced variable may be null

---

*Comment posted by [LGTM.com](https://lgtm.com)*

psiinon commented:
This pull request **introduces 9 alerts** when merging 9da329a6057d94debfffee336f445daa7aabbfc3 into de4537e305d018e8a185c0b7feb5f87576f021c9 - [view on LGTM.com](https://lgtm.com/projects/g/zaproxy/zap-extensions/rev/pr-bb732b4bd4d9490293d02495f60e073dd9bac5d1)

**new alerts:**

* 2 for Equals on incomparable types
* 2 for Implicit conversion from array to string
* 1 for Container contents are never accessed
* 1 for Inconsistent equals and hashCode
* 1 for Type mismatch on container modification
* 1 for Reference equality test of boxed types
* 1 for Dereferenced variable may be null

---

*Comment posted by [LGTM.com](https://lgtm.com)*

psiinon commented:
This pull request **introduces 9 alerts** when merging 4ec16be53b694cac55a7019c6b6ddf2be0eb74be into 0ceff13117188accffd473b5c36075b6d79ef55a - [view on LGTM.com](https://lgtm.com/projects/g/zaproxy/zap-extensions/rev/pr-b7e2c3465d52b5a69c5a5e21df40dd6632c04be5)

**new alerts:**

* 2 for Equals on incomparable types
* 2 for Implicit conversion from array to string
* 1 for Container contents are never accessed
* 1 for Inconsistent equals and hashCode
* 1 for Type mismatch on container modification
* 1 for Reference equality test of boxed types
* 1 for Dereferenced variable may be null

---

*Comment posted by [LGTM.com](https://lgtm.com)*

ManosMagnus commented:
Conflicts Resolved, and branch rebased on latest change of #1742

psiinon commented:
This pull request **introduces 4 alerts** when merging 99a97ca3d1d1e68e0d517dda9e99a906ba1c2042 into 0ceff13117188accffd473b5c36075b6d79ef55a - [view on LGTM.com](https://lgtm.com/projects/g/zaproxy/zap-extensions/rev/pr-cae06c800f91282ecc84a4fb5645a16b0780e39c)

**new alerts:**

* 1 for Container contents are never accessed
* 1 for Equals on incomparable types
* 1 for Type mismatch on container modification
* 1 for Expression always evaluates to the same value

---

*Comment posted by [LGTM.com](https://lgtm.com)*

thc202 commented:
The pull request is now updated (after build changes, zaproxy/zaproxy#5302).

eakirtas committed:
Websocket Gradle: Add GSON library

eakirtas committed:
[WS-Tree-Map] Adapt WebSocket Tree Map Structure to Payload Analyzer

eakirtas committed:
[WS-Tree-Map] Payload Analyzer Structure and JSON analyzer.

This commit aims to introduce a formal method to analyze and "represent"
websocket message payload. Every analyzer should implement the PayloadAnalyzer class
and produce a PayloadStructure. Every payload should be analyzed into
PayloadStructure.

TODO:
  [ ] Implement an API
  [ ] Add html Analyzer.
  [ ] Add methods in order to iterate on diferent paremeters

psiinon commented:
Thank you for this PR, and sorry it has taken us so long to address it.
We have decided to close it, but if you are interested in revisiting it, or in contributing to ZAP in any way then please get in touch.
